### PR TITLE
Use TaskDialog from Microsoft.WindowsAPICodePack

### PR DESCRIPTION
**Pull request by RussKie**
* Rework all callsites
* Remove PSTaskDialog project

This will make it easier to port to .NET 5.0

### CODE REVIEW

File: GitUI/CommandsDialogs/FormCommit.cs [hunk 1]
```diff
@@ -88,7 +89,9 @@ public sealed partial class FormCommit : GitModuleForm
         private readonly TranslationString _notOnBranch =
             new TranslationString("This commit will be unreferenced when switching to another branch and can be lost." +
                                   Environment.NewLine + Environment.NewLine + "Do you want to continue?");
-        private readonly TranslationString _notOnBranchButtons = new TranslationString("Checkout branch|Create branch|Continue");
+        private readonly TranslationString _notOnBranchButtonCheckout = new TranslationString("Checkout branch");
```
Comment (gerhardol) [inline GitUI/CommandsDialogs/FormCommit.cs:92]: Suggest moving to Strings.cs, shared with FormPull

File: GitUI/CommandsDialogs/FormStash.cs [hunk 2]
```diff
@@ -270,25 +271,28 @@ private void ClearClick(object sender, EventArgs e)
...
+                    TaskDialogResult result = dialog.Show();
+
+                    if (result == TaskDialogResult.Yes)
                     {
                         UICommands.StashDrop(this, stashName);
                         Initialize();
                     }
 
-                    if (PSTaskDialog.cTaskDialog.VerificationChecked)
+                    if (dialog.FooterCheckBoxChecked == true)
```
Comment (mstv) [inline GitUI/CommandsDialogs/FormStash.cs:295]: ditto

File: GitUI/CommandsDialogs/InvalidRepositoryRemover.cs [hunk 3]
```diff
@@ -22,37 +23,58 @@ public bool ShowDeleteInvalidRepositoryDialog(string repositoryPath)
         {
             int invalidPathCount = ThreadHelper.JoinableTaskFactory.Run(() => RepositoryHistoryManager.Locals.LoadRecentHistoryAsync())
                                                                    .Count(repo => !GitModule.IsValidGitWorkingDir(repo.Path));
-            string commandButtonCaptions = Strings.RemoveSelectedInvalidRepository;
+            int dialogResult = -1;
```
Comment (mstv) [inline GitUI/CommandsDialogs/InvalidRepositoryRemover.cs:26]: ditto

File: GitUI/MessageBoxes.cs [hunk 4]
```diff
@@ -85,21 +86,21 @@ public static bool CacheHostkey(IWin32Window owner)
...
+                Caption = Instance._updateSubmodules.Text,
+                Icon = TaskDialogStandardIcon.Information,
+                StandardButtons = TaskDialogStandardButtons.Yes | TaskDialogStandardButtons.No,
+                FooterCheckBoxText = Instance._rememberChoice.Text,
+                FooterIcon = TaskDialogStandardIcon.Information,
+                StartupLocation = TaskDialogStartupLocation.CenterOwner
+            };
+
+            bool result = dialog.Show() == TaskDialogResult.Yes;
+            if (dialog.FooterCheckBoxChecked == true)
```
Comment (mstv) [inline GitUI/MessageBoxes.cs:103]: ditto

File: GitUI/UserControls/RevisionGrid/RevisionGridControl.cs [hunk 5]
```diff
@@ -1837,71 +1838,78 @@ private string GetRefUnambiguousName(IGitRef gitRef)
...
+                StandardButtons = TaskDialogStandardButtons.Yes | TaskDialogStandardButtons.No,
+                Icon = TaskDialogStandardIcon.Information,
+                FooterCheckBoxText = _dontShowAgain.Text,
+                FooterIcon = TaskDialogStandardIcon.Information,
+                StartupLocation = TaskDialogStartupLocation.CenterOwner,
+            };
+
+            TaskDialogResult result = dialog.Show();
+
+            if (dialog.FooterCheckBoxChecked == true)
```
Comment (mstv) [inline GitUI/UserControls/RevisionGrid/RevisionGridControl.cs:1867]: ditto

File: GitUI/UserControls/RevisionGrid/RevisionGridControl.cs [hunk 6]
```diff
@@ -1837,71 +1838,78 @@ private string GetRefUnambiguousName(IGitRef gitRef)
...
```
Comment (mstv) [inline GitUI/UserControls/RevisionGrid/RevisionGridControl.cs:1905]: ditto

File: GitUI/CommandsDialogs/FormBrowse.cs [hunk 7]
```diff
@@ -1843,7 +1843,7 @@ private void FileExplorerToolStripMenuItemClick(object sender, EventArgs e)
         {
             try
             {
-                Process.Start(Module.WorkingDir);
+                Process.Start("explorer.exe", Module.WorkingDir);
```
[the remark this answers is not in the document]
Comment (RussKie) [reply GitUI/CommandsDialogs/FormBrowse.cs:1846]: Missed it, thaks.
This is a port from .NET Core port branch, not needed for net461

File: GitUI/CommandsDialogs/FormCheckoutBranch.cs [hunk 8]
```diff
@@ -381,19 +381,22 @@ private DialogResult OkClick()
...
+                            Caption = _applyStashedItemsAgainCaption.Text,
+                            Icon = TaskDialogStandardIcon.Information,
+                            StandardButtons = TaskDialogStandardButtons.Yes | TaskDialogStandardButtons.No,
+                            FooterCheckBoxText = _dontShowAgain.Text,
+                            FooterIcon = TaskDialogStandardIcon.Information,
+                            StartupLocation = TaskDialogStartupLocation.CenterOwner
+                        };
+
+                        TaskDialogResult result = dialog.Show();
+                        messageBoxResult = result == TaskDialogResult.Yes;
```
[the remark this answers is not in the document]
Comment (RussKie) [reply GitUI/CommandsDialogs/FormCheckoutBranch.cs:397]: No reason

File: GitExtensions/Program.cs [hunk 9]
```diff
@@ -295,16 +296,31 @@ private static void HandleConfigurationException(ConfigurationException ce)
...
-                                                                        MainInstruction: ResourceManager.Strings.GitExecutableNotFound,
-                                                                        Content: null,
-                                                                        ExpandedInfo: null,
-                                                                        Footer: null,
-                                                                        VerificationText: null,
-                                                                        CommandButtons: $"{ResourceManager.Strings.FindGitExecutable}|{ResourceManager.Strings.InstallGitInstructions}",
-                                                                        ShowCancelButton: true,
-                                                                        MainIcon: PSTaskDialog.eSysIcons.Error,
-                                                                        FooterIcon: PSTaskDialog.eSysIcons.Warning);
+            int dialogResult = -1;
```
[the remark this answers is not in the document]
Comment (RussKie) [reply GitExtensions/Program.cs:299]: Values are arbitrary, so enum won't work.

> Or rather move the handling code in the `switch` to the `Click` handlers, perhaps using local functions.

Feel free to offer an alternative.
Mind you in .NET 5.0 we have slightly different API, so being too clever now **will* inhibit the future port.

File: GitUI/CommandsDialogs/FormPull.cs [hunk 10]
```diff
@@ -544,19 +566,21 @@ void PopStash()
...
+                        StandardButtons = TaskDialogStandardButtons.Yes | TaskDialogStandardButtons.No,
+                        Icon = TaskDialogStandardIcon.Information,
+                        FooterCheckBoxText = _dontShowAgain.Text,
+                        FooterIcon = TaskDialogStandardIcon.Information,
+                        StartupLocation = TaskDialogStartupLocation.CenterOwner,
+                    };
+
+                    messageBoxResult = dialog.Show() == TaskDialogResult.Yes;
+
+                    if (dialog.FooterCheckBoxChecked == true)
```
[the remark this answers is not in the document]
Comment (RussKie) [reply GitUI/CommandsDialogs/FormPull.cs:583]: This is a nullable property, so it won't compile.

File: GitUI/CommandsDialogs/FormBrowse.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (mstv) [reply GitUI/CommandsDialogs/FormBrowse.cs:1846]: So `Process` will not support `ProcessStartInfo.UseShellExecute` anymore?!?

File: GitExtensions/Program.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (mstv) [reply GitExtensions/Program.cs]: Any objections against 48ad06449fe2ed87173fa457f2cb729515b608aa?
If not, I'll adapt the other occurrances.

[the remark this answers is not in the document]
Comment (RussKie) [reply GitExtensions/Program.cs]: That would work, though I don't quite like local (especially nested) functions.
Please have a look at the .NET 5.0 TaskDialog sample - https://github.com/dotnet/samples/pull/3702/files#diff-3be29162f7cb32efc1b79353c60c16dc.

In .NET 5.0 version the code would be something like:
```
            TaskDialogButton btnCancel = ...
            TaskDialogButton btnSave = ...
            TaskDialogButton btnDontSave = ...

            TaskDialogButton result = TaskDialog.ShowDialog(this, page);

            if (result == btnSave)
                Console.WriteLine("Saving");
            else if (result == btnDontSave)
                Console.WriteLine("Not saving");
            else
                Console.WriteLine("Canceling");
```

File: GitUI/CommandsDialogs/FormBrowse.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (RussKie) [reply GitUI/CommandsDialogs/FormBrowse.cs:1846]: On .NET Core branch the original code didn't work.

File: GitExtensions/Program.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (mstv) [reply GitExtensions/Program.cs]: OK, so with .NET 5.0 these ugly magic numbers will be removed. Then I withdraw my commit.

File: GitUI/CommandsDialogs/FormBrowse.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (mstv) [reply GitUI/CommandsDialogs/FormBrowse.cs:1846]: So we can keep it here already.

**Comment by RussKie** [comment]
I'm going to squash-merge tomorrow

**Pull request merged**